server/api/admin.go

```diff
@@ -127,3 +150,78 @@ func (h *adminHandler) SavePersistFile(w http.ResponseWriter, r *http.Request) {
 	}
 	h.rd.Text(w, http.StatusOK, "")
 }
+
+func (h *adminHandler) MarkRecovering(w http.ResponseWriter, r *http.Request) {
+	if err := h.svr.MarkRecovering(); err != nil {
+		_ = h.rd.Text(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+	_ = h.rd.Text(w, http.StatusOK, "")
+}
+
+func (h *adminHandler) IsRecoveringMarked(w http.ResponseWriter, r *http.Request) {
+	marked, err := h.svr.IsRecoveringMarked(r.Context())
+	if err != nil {
+		_ = h.rd.Text(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+	type resStruct struct {
+		Marked bool `json:"marked"`
+	}
+	_ = h.rd.JSON(w, http.StatusOK, &resStruct{Marked: marked})
+}
+
+func (h *adminHandler) UnmarkRecovering(w http.ResponseWriter, r *http.Request) {
+	if err := h.svr.UnmarkRecovering(r.Context()); err != nil {
+		_ = h.rd.Text(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+	_ = h.rd.Text(w, http.StatusOK, "")
+}
+
+// RecoverAllocID recover base alloc id
+// body should be in {"id": "123"} format
+func (h *adminHandler) RecoverAllocID(w http.ResponseWriter, r *http.Request) {
+	var input map[string]interface{}
+	if err := apiutil.ReadJSONRespondError(h.rd, w, r.Body, &input); err != nil {
+		return
+	}
+	idValue, ok := input["id"].(string)
+	if !ok || len(idValue) == 0 {
+		_ = h.rd.Text(w, http.StatusBadRequest, "invalid id value")
+		return
+	}
+	newId, err := strconv.ParseUint(idValue, 10, 64)
+	if err != nil {
+		_ = h.rd.Text(w, http.StatusBadRequest, err.Error())
+		return
+	}
+	marked, err := h.svr.IsRecoveringMarked(r.Context())
+	if err != nil {
+		_ = h.rd.Text(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+	if !marked {
+		_ = h.rd.Text(w, http.StatusForbidden, "can only recover alloc id when recovering mark marked")
+		return
+	}
+
+	leader := h.svr.GetLeader()
+	if leader == nil {
+		_ = h.rd.Text(w, http.StatusServiceUnavailable, errs.ErrLeaderNil.FastGenByArgs().Error())
+		return
+	}
+	ctx := grpcutil.BuildForwardContext(r.Context(), leader.ClientUrls[0])
```

Comment:
Why does it create a gRPC request in HTTP API?

Reply:
only leader can set it, so make it an rpc. http wraps it.

Reply:
The HTTP request will redirect automatically, see [server/api/server.go](https://github.com/tikv/pd/blob/6d883d0e8386c64ac12ade25d00893a608b33a14/server/api/server.go#L39)